include/robot_interfaces/logger.hpp

```diff
@@ -0,0 +1,287 @@
+    {
+        thread_ = std::make_shared<real_time_tools::RealTimeThread>();
+    }
+
+    virtual ~FingerLogger()
+    {
+        stop_was_called_ = true;
+        stop();
+        thread_->join();
+
+    }
+
+    void get_header()
+    {
+      //using the newest index inofrmation in logger_data to extract information on the fields to be logged
+      applied_action = (*logger_data_->applied_action)[logger_data_->applied_action->newest_timeindex()];
+      desired_action = (*logger_data_->desired_action)[logger_data_->desired_action->newest_timeindex()];
+      observation = (*logger_data_->observation)[logger_data_->observation->newest_timeindex()];
+      status = (*logger_data_->status)[logger_data_->status->newest_timeindex()];
+
+      header.push_back("#");
+      header.push_back("[Timestamp]");
+      header.push_back("(Current Index)");
+
+      for(i = 0; i < observation_name_.size(); i++)
+      {
+        for(j = 0; j < observation_data_[i].size(); j ++)
+        {
+          temp = observation_name_[i] + " " + std::to_string(j);
+          header.push_back(temp);
+        }
+      }
+
+      for(i = 0; i < applied_action_name_.size(); i++)
+      {
+        for(j = 0; j < applied_action_data_[i].size(); j ++)
+        {
+          temp = applied_action_name_[i] + " " + std::to_string(j);
+          header.push_back(temp);
+        }
+      }
+
+      for(i = 0; i < desired_action_name_.size(); i++)
+      {
+        for(j = 0; j < desired_action_data_[i].size(); j ++)
+        {
+          temp = desired_action_name_[i] + " " + std::to_string(j);
+          header.push_back(temp);
+        }
+      }
+
+      header.push_back("Status");
+
+    }
+
+    static void *write(void *instance_pointer)
+    {
+        ((FingerLogger *)(instance_pointer))->write();
+        return nullptr;
+    }
+
+    void write()
+    {
+
+        output_file_.open(output_file_name_);
+        std::ostream_iterator<std::string> output_iterator(output_file_, " , ");
+        get_header();
+        std::copy(header.begin(), header.end(), output_iterator);
+        output_file_ << std::endl;
+        output_file_.close();
+
+        while (!stop_was_called_ &&
+               !(logger_data_->desired_action->length() > 0))
+        {
+            real_time_tools::Timer::sleep_until_sec(0.1);
+        }
+
+        index_ = logger_data_->observation->newest_timeindex();
+
+        for (long int t = 0; !stop_was_called_; t++)
+        {
+            if (index_ + block_size_ ==
+                logger_data_->observation->newest_timeindex())
+            {
+                output_file_.open(output_file_name_, std::ios_base::app);
+
+
+#ifdef VERBOSE
+                auto t1 = std::chrono::high_resolution_clock::now();
+#endif
+
+                for (long int j = index_; j < index_ + block_size_; j++)
+                {
+                    try
+                    {
+                        std::ostream_iterator<double> output_iterator_(output_file_, " , ");
```

Comment:
Did you find a solution for the trailing comma at the end of the line?

Reply:
So, I have used an extremely hacky solution rn, just so that it works. I am basically using the same ostream_iterator only till the second last value. Then using another ostream_iterator with newline as the separator. Looking for a better solution.

Reply:
I just found this [infix_iterator](https://stackoverflow.com/a/3497021/2095383) which is supposed to be a direct replacement for `std::ostream_iterator` without the trailing separator.
We could add this to the `mpi_cpp_tools` package.